awx/main/utils/common.py

```diff
@@ -531,7 +531,16 @@ def copy_m2m_relationships(obj1, obj2, fields, kwargs=None):
                 src_field_value = getattr(obj1, field_name)
                 if kwargs and field_name in kwargs:
                     override_field_val = kwargs[field_name]
+                    if field_name == 'instance_groups':
+                        # instance_groups are a list but we need to preserve the order
+                        for ig_id in override_field_val:
+                            getattr(obj2, field_name).add(ig_id)
+                        continue
                     if isinstance(override_field_val, (set, list, QuerySet)):
+                        # Labels are additive so we are going to add any src labels in addition to the override labels
+                        if field_name == 'labels':
+                            for jt_label in src_field_value.all():
+                                getattr(obj2, field_name).add(jt_label.id)
```

Comment:
Merging of credentials in this process is handled here:

https://github.com/ansible/awx/blob/063b0d08343938295a3c8f6ccb5342ffd2421d40/awx/main/models/unified_jobs.py#L402-L409

Yes, all 3 relationships, instance_groups, labels, and credentials have distinct rules associated with them. But handing 1 relationship _before_ the method `copy_m2m_relationships` call, and 2 relationships _inside_ the method is going to make this much harder for anyone to ever follow in the future.

To pick one, I would say it should be done before the method call, because can be adjusted to be more testable. That still leaves you with an exception case for _saving_ instance groups, which would have to remain here.